src/mscorlib/shared/System/IO/PathInternal.Windows.cs

```diff
@@ -367,12 +436,13 @@ internal static string NormalizeDirectorySeparators(string path)
                 if (normalized) return path;
             }
 
-            StringBuilder builder = new StringBuilder(path.Length);
-
+            Span<char> result = new char[path.Length];
```

Comment:
Note here you just made an allocation on the heap (new array) which is what we're trying to avoid.
We do need an allocation on the heap (since we don't know the size of path at compile time we can't use stackalloc etc) but ideally it would be a pooled one.
Take a look at the implementation of ValueStringBuilder. When it runs out of space, it uses ArrayPool<char> to borrow a premade one. If you initialize it with ReadOnlySpan<char>.Empty then on the first Append it will go get one of those.
Unfortunately that still isn't great as it doesn't know that it should try to get one at least as long as path. To fix that I would add another constructor to ValueStringBuilder that takes an int capacity (setting the buffer to ReadOnlySpan<char>.Empty). On the first call to Grow, it will grow to at least that capacity. 
Copy your changes to VSB into the copy in corefx\src\Common\src\System\Text\ValueStringBuilder.cs and update the tests for that in CoreFX.

Reply:
your goal is that the new methods you added do not contain the `new` keyword in any place or anything they call into except error cases and for newing structs (like Span). that will eliminate allocation on the heap

Reply:
> add another constructor to `ValueStringBuilder` that takes an int capacity

I do not think we should be adding more fields to `ValueStringBuilder`.

> your goal is that the new methods you added do not contain the new keyword in any place

This is not actually possible. These methods have to allocate in cases where the returned string is not a substring of the original `ReadOnlySnap<char>`.

Reply:
@jkotas yes obviously, I was imprecise.

> I do not think we should be adding more fields to ValueStringBuilder.

What about adding such a constructor, which immediately calls Grow with the provided value? No new field, and it avoids this allocation.

Reply:
That's fine.